pkg/exemplars/tsdb.go

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package exemplars
+
+import (
+	"github.com/gogo/status"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/storage"
+	"google.golang.org/grpc/codes"
+
+	"github.com/thanos-io/thanos/pkg/exemplars/exemplarspb"
+	"github.com/thanos-io/thanos/pkg/store/labelpb"
+)
+
+// TSDB implements exemplarspb.Exemplars that allows to fetch exemplars from a TSDB instance.
+type TSDB struct {
+	db        storage.ExemplarQueryable
+	extLabels labels.Labels
+}
+
+// NewTSDB creates new exemplars.TSDB.
+func NewTSDB(db storage.ExemplarQueryable, extLabels labels.Labels) *TSDB {
+	return &TSDB{
+		db:        db,
+		extLabels: extLabels,
+	}
+}
+
+// Exemplars returns all specified exemplars from a TSDB instance.
+func (t *TSDB) Exemplars(r *exemplarspb.ExemplarsRequest, s exemplarspb.Exemplars_ExemplarsServer) error {
+	expr, err := parser.ParseExpr(r.Query)
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
+
+	match, selectors := selectorsMatchesExternalLabels(parser.ExtractSelectors(expr), t.extLabels)
+
+	if !match {
+		return nil
+	}
+
+	if len(selectors) == 0 {
+		return status.Error(codes.InvalidArgument, errors.New("no matchers specified (excluding external labels)").Error())
+	}
```

Comment:
Line 41 ~ 47 looks very nice! Would you mind also updating https://github.com/thanos-io/thanos/blob/main/pkg/exemplars/proxy.go#L67-L76? 
It is good to return an error if no other matcher specified instead of returning nil.

Reply:
Done